src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -73,22 +71,11 @@ public static object Deserialize(string json, Type returnType, JsonSerializerOpt
             object result;
             byte[] tempArray = null;
 
-            int maxBytes;
-
-            // For performance, avoid asking for the actual byte count unless necessary.
-            if (json.Length > MaxArrayLengthBeforeCalculatingSize)
-            {
-                // Get the actual byte count in order to handle large input.
-                maxBytes = JsonReaderHelper.GetUtf8ByteCount(json.AsSpan());
-            }
-            else
-            {
-                maxBytes = json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
-            }
-
-            Span<byte> utf8 = maxBytes <= JsonConstants.StackallocThreshold ?
-                stackalloc byte[maxBytes] :
-                (tempArray = ArrayPool<byte>.Shared.Rent(maxBytes));
+            // For performance, avoid obtaining the actual byte count unless the memory usage may be
+            // higher than the threshold which is options.DefaultBufferSize.
+            Span<byte> utf8 = json.Length <= (ArrayPoolMaxSizeBeforeUsingNormalAlloc / JsonConstants.MaxExpansionFactorWhileTranscoding) ?
+                tempArray = ArrayPool<byte>.Shared.Rent(json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding) :
+                new byte[JsonReaderHelper.GetUtf8ByteCount(json.AsSpan())];
```

Comment:
Why not just always use Rent?  If it's too large for the pool, it'll just allocate it itself.

Reply:
From the previous commit (which removed the comment):
https://github.com/dotnet/corefx/pull/41414/files/a8cd02e62d6bd2d4686ff86124fdc31e233ca42d..218e1c0d306dde4441927fdcecacd071295a145c

> // and because we can avoid calling Clear().

And from above:

> // For performance, avoid obtaining the actual byte count unless the memory usage may be
// higher than the threshold

Reply:
> avoid obtaining the actual byte count unless the memory usage may be higher than the threshold

Computing an accurate byte count doesn't prevent you from still using ArrayPool.

> and because we can avoid calling Clear()

The benefit analysis of clearing all of these arrays is still not obvious to me.

Reply:
> Computing an accurate byte count doesn't prevent you from still using ArrayPool.

I believe the intention here is to avoid doing the extra work to get an accurate byte count when we are using the arraypool anyway and save that cost (asking for 4k from the pool vs 12k isn't that much different, so asking for the worst case is good enough). However, when allocating a regular array, over-asking by up to 3x is probably too expensive, so asking for the exact amount needed by incurring the cost of getting the exact count is worth the trade-off.

With your suggestion, we'd always have to always get the accurate byte count, regardless of whether the pool was allocating itself (because it got exhausted/we need over 1 MB), or not. Maybe that approach is good enough but there are certainly some trade-offs here. @steveharter, what are your thoughts on this?

Reply:
Yes there are two things going on here:
- Avoid asking for a transcoded byte count when a threshold is surpassed
- The threshold selected happens to be the same as ArrayPool's current implementation where it just does a normal alloc, so a call to Clear() can be avoided

Given that, yes the call to Clear() is insignificant so we could just use ArrayPool always.